libraries/AP_Motors/AP_MotorsMatrixSlewRateLimited.cpp

```diff
@@ -0,0 +1,300 @@
+       failed = true;
+    }
+
+    _last_highest_motor = highest_motor_index;
+
+    return failed;
+}
+
+/**
+ * Start a motor recovery
+ */
+void AP_MotorsMatrixSlewRateLimited::start_motor_recovery(uint64_t current_time)
+{
+    set_slew_rate(_params._recovery_slew_rate);
+
+    // Set the last motor output based on the recovery percentage
+    // The slew rate limiter will force the actual outputs to be this
+    // value +- one slew rate increment during the next loop iteration.
+    float recovery_pwm = out_to_pwm(_params._recovery_output) - _slew_per_iteration;
+    for (uint8_t i=0; i<AP_MOTORS_MAX_NUM_MOTORS; i++) {
+        if (motor_enabled[i]) {
+            _motor_out_pwm_max[i] = recovery_pwm;
+        }
+    }
+
+    // calculate the total recovery time, in seconds
+    float recovery_time_length = (get_pwm_output_max() - out_to_pwm(_params._recovery_output)) / _params._recovery_slew_rate;
+    // set the state transition time after recovery is complete
+    _next_state_transition_time = current_time + (recovery_time_length * 1000 * 1000);
+}
+
+/**
+ * Get the motor output as a percentage
+ */
+float AP_MotorsMatrixSlewRateLimited::get_requested_motor_out(uint8_t mot)
+{
+    return (mot<AP_MOTORS_MAX_NUM_MOTORS) ? pwm_to_out(_motor_out_pwm_requested[mot]) : 0.0f;
+}
+
+/**
+ * Convert a PWM value to a proportion of full output, using the
+ * configured pwm_output_max and pwm_output_min.
+ */
+float AP_MotorsMatrixSlewRateLimited::pwm_to_out(uint16_t pwm)
+{
+    return ((float)pwm - get_pwm_output_min()) / (get_pwm_output_max()-get_pwm_output_min());
+}
+
+/**
+ * Convert a proportion of full output to a PWM value, using the
+ * configured pwm_output_max and pwm_output_min.
+ */
+float AP_MotorsMatrixSlewRateLimited::out_to_pwm(float out)
+{
+    return (get_pwm_output_min() + ((get_pwm_output_max()-get_pwm_output_min()) * out));
+}
+
+/**
+ * Set the slew per iteration value based on a slew rate in slew per seconds
+ */
+void AP_MotorsMatrixSlewRateLimited::set_slew_rate(float slew_per_second)
+{
+    _slew_per_iteration = slew_per_second / _loop_rate;
+}
+
+/**
+ * Clears / resets the motor failure detection state
+ */
+void AP_MotorsMatrixSlewRateLimited::reset_failure_detector(uint64_t current_time)
+{
+    _last_motor_good_time = current_time;
+    for (uint8_t i=0; i<AP_MOTORS_MAX_NUM_MOTORS; i++) {
+        _motor_out_filtered[i] = get_requested_motor_out(i);
+    }
+}
+
+void AP_MotorsMatrixSlewRateLimited::set_state(motor_recovery_state new_state)
+{
+    _recovery_state = new_state;
+    log_state(true);
+}
+
+void AP_MotorsMatrixSlewRateLimited::log_state(bool force)
+{
+    uint64_t current_time = AP_HAL::micros64();
+
+    if (force || current_time > _next_log_time) {
+        _next_log_time = current_time + (_params._log_interval_ms * 1000);
+
+        DataFlash_Class::instance()->Log_Write("SLW1",
+            "TimeUS,State,FailCnt,NxtTrans,SlwRt", "QIHIf",
```

Comment:
I think this one is missing units and multipliers information

Reply:
Good point. These are going to get removed / reworked when I make the changes to remove the ESC failure detection code. If any of these logging statements remain, I'll make sure to include units.